pkgs/native_assets_cli/lib/src/config.dart

```diff
@@ -0,0 +1,538 @@
+import 'build_mode.dart';
+import 'encoded_asset.dart';
+import 'json_utils.dart';
+import 'model/metadata.dart';
+import 'os.dart';
+import 'utils/datetime.dart';
+import 'utils/json.dart';
+
+/// The shared properties of a [LinkConfig] and a [BuildConfig].
+///
+/// This abstraction makes it easier to design APIs intended for both kinds of
+/// build hooks, building and linking.
+sealed class HookConfig {
+  /// The underlying json configuration of this [HookConfig].
+  final Map<String, Object?> json;
+
+  /// The version of the [HookConfig].
+  final Version version;
+
+  /// The directory in which output and intermediate artifacts that are unique
+  /// to this configuration can be placed.
+  ///
+  /// This directory is unique per hook and per configuration.
+  ///
+  /// The contents of this directory will not be modified by anything else than
+  /// the hook itself.
+  ///
+  /// The invoker of the the hook will ensure concurrent invocations wait on
+  /// each other.
+  final Uri outputDirectory;
+
+  /// The directory in which shared output and intermediate artifacts can be
+  /// placed.
+  ///
+  /// This directory is unique per hook.
+  ///
+  /// The contents of this directory will not be modified by anything else than
+  /// the hook itself.
+  ///
+  /// The invoker of the the hook will ensure concurrent invocations wait on
+  /// each other.
+  final Uri outputDirectoryShared;
+
+  /// The name of the package the assets are built for.
+  final String packageName;
+
+  /// The root of the package the assets are built for.
+  ///
+  /// Often a package's assets are built because a package is a dependency of
+  /// another. For this it is convenient to know the packageRoot.
+  final Uri packageRoot;
+
+  /// The operating system being compiled for.
+  final OS targetOS;
+
+  /// The [BuildMode] that the code should be compiled in.
+  ///
+  /// Currently [BuildMode.debug] and [BuildMode.release] are the only modes.
+  ///
+  /// Not available during a dry run.
+  final BuildMode? _buildMode;
+
+  /// The asset types that the invoker of this hook supports.
+  final List<String> supportedAssetTypes;
+
+  HookConfig(this.json)
+      : version = switch (Version.parse(json.string(_versionKey))) {
+          final Version version => (version.major != currentVersion.major ||
+                  version < currentVersion)
+              ? throw FormatException(
+                  'Only compatible versions with $currentVersion are supported '
+                  '(was: $version).')
+              : version,
+        },
+        outputDirectory = json.path(_outDirConfigKey),
+        outputDirectoryShared = json.path(_outDirSharedConfigKey),
+        packageRoot = json.path(_packageRootConfigKey),
+        packageName = json.string(_packageNameConfigKey),
+        targetOS = OS.fromString(json.string(_targetOSConfigKey)),
+        supportedAssetTypes =
+            json.optionalStringList(_supportedAssetTypesKey) ?? const [],
```

Comment:
This is not the right default. The right default for older SDKs is CodeAssets.
However, it should always be passed as this param was added in protocol version 1.1.
So this should be non-optional string list instead.

Reply:
[Already the previous CL ](https://github.com/dart-lang/native/commit/d8a84ead5ecd85a08fb982cea00cc400caa1ebae) required a bundling tool to be explicit about the supported assets. If it doesn't supply any it's empty. We shouldn't silently assume that a bundling tool supports code assets - it may not support it.

We should go through the json writers to encode things no longer in an optional way (i.e. empty list is written as empty list not as `null`) and make readers mandate them.